File: PYTHON/Algorithms/Binary_search.py
from typing import List

# recursion
# Time : O(log N)
# Space : O(log N)
def BinarySearch_recursion(nums: List[int], target: int) -> int:
    return recursion_helper(nums, 0, len(nums) - 1, target)

def recursion_helper(nums, left, right, target):
    if left > right:
        return -1

    mid = (left + right) // 2

    if target == nums[mid]:
        return mid
    elif target < nums[mid]:
        return recursion_helper(nums, left, mid - 1, target)
    else:
        return recursion_helper(nums, mid + 1, right, target)

# while loop
# Time : O(log N)
# Space : O(1)
def BinarySearch(nums: List[int], target: int) -> int:
    left = 0
    right = len(nums) - 1

    while left <= right:
        mid = (left + right) // 2
        if target == nums[mid]:
            return mid

        if target < nums[mid]:
            right = mid - 1
        else:
            left = mid + 1

    return -1


# Test for recursion
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9]
target_recursion = 6
result_recursion = BinarySearch_recursion(nums, target_recursion)
print(f"Recursion: Index of {target_recursion} is {result_recursion}")

# Test for while loop
target_while_loop = 3
result_while_loop = BinarySearch(nums, target_while_loop)
print(f"While Loop: Index of {target_while_loop} is {result_while_loop}")

# Test for a target not in the array
target_not_in_array = 10
result_recursion = BinarySearch_recursion(nums, target_not_in_array)
result_while_loop = BinarySearch(nums, target_not_in_array)
print(f"Recursion: Index of {target_not_in_array} is {result_recursion}")
print(f"While Loop: Index of {target_not_in_array} is {result_while_loop}")

File: JAVASCRIPT/Algorithms/Search/BinarySearch.js
const BinarySearch = (sortedNums, target) => {
  let left = 0;
  let right = sortedNums.length - 1;

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    if (sortedNums[mid] === target) {
      return mid;
    } else if (sortedNums[mid] < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
};

const BinarySearchRecursion = (sortedNums, target) => {
  return recursionHelper(sortedNums, 0, sortedNums.length - 1, target);
};

const recursionHelper = (sortedNums, left, right, target) => {
  if (left > right) {
    return -1;
  }
  const mid = Math.floor((left + right) / 2);

  if (sortedNums[mid] === target) {
    return mid;
  } else if (sortedNums[mid] < target) {
    return recursionHelper(sortedNums, mid + 1, right, target);
  } else {
    return recursionHelper(sortedNums, left, mid - 1, target);
  }
};

const nums = [2, 4, 5, 17, 28, 49];

// Test Recursion
const recursionTarget = 5;
const recursionResult = BinarySearchRecursion(nums, recursionTarget);
console.log(`Recursion target ${recursionTarget} found on ${recursionResult}th index`);

// Test While Loop
const loopTarget = 49;
const loopResult = BinarySearch(nums, loopTarget);
console.log(`Loop target ${loopTarget} found on ${loopResult}th index`);

// Test Target not in Nums
const targetNotInNums = 100;
const loopResultNotInNums = BinarySearch(nums, targetNotInNums);
const recursionResultNotInNums = BinarySearchRecursion(nums, targetNotInNums);
console.log(`index of ${targetNotInNums} in nums array is ${loopResultNotInNums}`);
